pkg/pgmodel/cache/cache.go

```diff
@@ -51,26 +56,68 @@ type MetricNameCache struct {
 	Metrics *clockcache.Cache
 }
 
+func (m *MetricNameCache) key(schema, metric string) (int, string) {
+	// deep copy the strings so the original memory doesn't need to stick around
+	keyBuilder := strings.Builder{}
+	keyBuilder.Grow(len(metric) + len(schema) + len(fieldSeparator))
+	keyBuilder.WriteString(schema)
+	keyBuilder.WriteString(fieldSeparator)
+	keyBuilder.WriteString(metric)
+	return keyBuilder.Len(), keyBuilder.String()
+}
+
+func (m *MetricNameCache) value(schema, tableName, seriesTable string) (int, string) {
+	// deep copy the strings so the original memory doesn't need to stick around
+	valueBuilder := strings.Builder{}
+	valueBuilder.Grow(len(schema) + len(tableName) + len(seriesTable) + (len(fieldSeparator) * 2))
+	valueBuilder.WriteString(schema)
+	valueBuilder.WriteString(fieldSeparator)
+	valueBuilder.WriteString(tableName)
+	valueBuilder.WriteString(fieldSeparator)
+	valueBuilder.WriteString(seriesTable)
+	return valueBuilder.Len(), valueBuilder.String()
+}
+
+func (m *MetricNameCache) fromValue(val string) (string, string, string, error) {
+	values := strings.Split(val, fieldSeparator)
+
+	if len(values) != 3 {
+		return "", "", "", fmt.Errorf("invalid number of values stored")
+	}
+
+	return values[0], values[1], values[2], nil
+}
+
 // Get fetches the table name for specified metric.
-func (m *MetricNameCache) Get(metric string) (string, error) {
-	result, ok := m.Metrics.Get(metric)
+func (m *MetricNameCache) Get(schema, metric string) (string, string, string, error) {
+	if schema == "" {
+		schema = nullSchemaName
+	}
+	_, key := m.key(schema, metric)
+	result, ok := m.Metrics.Get(key)
 	if !ok {
-		return "", errors.ErrEntryNotFound
+		return "", "", "", errors.ErrEntryNotFound
 	}
-	return result.(string), nil
+	return m.fromValue(result.(string))
+}
+
+// Set stores metric info for specified metric with schema.
+func (m *MetricNameCache) Set(schema, metric, tableName, seriesTable string) error {
+	// deep copy the strings so the original memory doesn't need to stick around
+	valLen, val := m.value(schema, tableName, seriesTable)
+	keyLen, key := m.key(schema, metric)
+	//size includes an 8-byte overhead for each string
+	m.Metrics.Insert(key, val, uint64(keyLen+valLen+16))
+	return nil
 }
 
-// Set stores table name for specified metric.
-func (m *MetricNameCache) Set(metric string, tableName string) error {
+// Set stores metric info cache for specified metric with default schema.
+func (m *MetricNameCache) SetDefaultSchema(schema, metric, tableName, seriesTable string) error {
```

Comment:
Can we define what a default-schema is, here? I assume it's something that will always be static which in that case, why do we accept a `schema` as param if it's gonna be the default?

Reply:
Default schema is the schema which is used if there is no schema specified. Should be determined by the DB state (i.e. if there is only one metric in the system with the name, its schema will be the default schema)

Reply:
I'm confused why is schema than an argument to this function?

Reply:
also how is this different than the logic on line 94? with nullSchemaName

Reply:
Updated to use single Set function.